src/main/java/net/dries007/tfc/common/recipes/SimpleCraftMatrix.java

```diff
@@ -0,0 +1,270 @@
+     *
+     * @param other Another craft matrix
+     * @return if 'other' is a subset of the current craft matrix (i.e. other is found somewhere within the current matrix)
+     */
+    public boolean matches(SimpleCraftMatrix other)
+    {
+        // Check all possible shifted positions
+        for (int xShift = 0; xShift <= this.width - other.width; xShift++)
+        {
+            for (int yShift = 0; yShift <= this.height - other.height; yShift++)
+            {
+                if (matches(other, xShift, yShift, false))
+                {
+                    return true;
+                }
+                else if (matches(other, xShift, yShift, true))
+                {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private boolean matches(SimpleCraftMatrix other, int startX, int startY, boolean isMirrored)
+    {
+        for (int x = 0; x < this.width; x++)
+        {
+            for (int y = 0; y < this.height; y++)
+            {
+                int patternIdx = y * width + x;
+                if (x < startX || y < startY || x - startX >= other.width || y - startY >= other.height)
+                {
+                    // If the current position in the matrix is outside the pattern, the value should be set by other.outsideSlot
+                    if (matrix[patternIdx] != other.outsideSlot)
+                        return false;
+                }
+                else
+                {
+                    // Otherwise, the value must equal the value in the pattern
+                    int otherIdx;
+                    if (isMirrored)
+                        otherIdx = (y - startY) * other.width + (other.width - 1 - (x - startX));
+                    else
+                        otherIdx = (y - startY) * other.width + (x - startX);
+
+                    if (matrix[patternIdx] != other.matrix[otherIdx])
+                        return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    public void toNetwork(PacketBuffer buffer, int width, int height)
+    {
+        buffer.writeVarInt(width);
+        buffer.writeVarInt(height);
+        short packed = 0;
+        for (int i = 0; i < matrix.length; i++)
+            if (matrix[i]) packed |= (1 << i);
+        buffer.writeShort(packed);
+    }
+
+    public static SimpleCraftMatrix fromNetwork(PacketBuffer buffer, boolean outsideSlotRequired)
+    {
+        int width = buffer.readVarInt();
+        int height = buffer.readVarInt();
+        boolean[] matrix = new boolean[width * height];
+
+        short packed = buffer.readShort();
+        for (int i = 0; i < matrix.length; i++)
+            matrix[i] = (packed & (1 << i)) != 0;
+        return new SimpleCraftMatrix(outsideSlotRequired, matrix, width, height);
+    }
+
+    /**
+     * From {@link net.minecraft.item.crafting.ShapedRecipe}
+     */
+    public static String[] patternFromJson(JsonArray array)
```

Comment:
Rather than deserializing `JsonArray` -> `String[]` -> `SimpleCraftMatrix` It'd be simpler just to have `fromJson` and return the matrix itself. i.e. a json object:
```json
{ "pattern": [...], "invert_empty_slots": true }
```

(as an aside I'm trying to think of better names for some of this stuff esp. since it's becoming serializable in json and yes I know I named them like two years ago yes they're crap)